src/snowflake/connector/connection.py

```diff
@@ -1541,13 +1543,27 @@ def is_an_error(status: QueryStatus) -> bool:
 
     def _all_async_queries_finished(self) -> bool:
         """Checks whether all async queries started by this Connection have finished executing."""
-        queries = copy.copy(
-            self._async_sfqids
-        )  # get_query_status might update _async_sfqids, let's copy the list
-        finished_async_queries = (
-            not self.is_still_running(self.get_query_status(q)) for q in queries
-        )
-        return all(finished_async_queries)
+
+        queries = list(reversed(list(self._async_sfqids.keys())))
+        if not queries:
+            return True
+        num_workers = min(self.client_prefetch_threads, len(queries))
+        found_unfinished_query = False
+
+        def check_status_helper(
+            sfq_id: str,
+        ):  # We should upgrade to using cancel_futures=True once supporting 3.9+
+            nonlocal found_unfinished_query
+            if not found_unfinished_query and self.is_still_running(
+                self.get_query_status(sfq_id)
+            ):
+                found_unfinished_query = True
+
+        with ThreadPoolExecutor(max_workers=num_workers) as tpe:
```

Comment:
Thank you for the improvements here!

Minor ask: For easier troubleshooting through logging, could an appropriately named `thread_name_prefix` be passed to this constructor to differentiate it from other thread-pools at runtime (such as those that download result batches, customer's own thread-pools, etc.)